File: fixtures/scale/java/medium/src/calc/eval/Evaluator.java
package calc.eval;

import calc.ast.AssignExpr;
import calc.ast.BinaryExpr;
import calc.ast.CallExpr;
import calc.ast.Expr;
import calc.ast.NumberExpr;
import calc.ast.UnaryExpr;
import calc.ast.VariableExpr;
import calc.errors.EvalException;
import calc.parser.OperatorTable;

public class Evaluator {
    private final Environment environment;
    private final FunctionRegistry registry;

    public Evaluator(Environment environment, FunctionRegistry registry) {
        this.environment = environment;
        this.registry = registry;
    }

    public double evaluate(Expr expr) {
        if (expr instanceof NumberExpr n) return n.getValue();
        if (expr instanceof VariableExpr v) return environment.lookup(v.getName());
        if (expr instanceof AssignExpr a) {
            double value = evaluate(a.getValue());
            environment.define(a.getName(), value);
            return value;
        }
        if (expr instanceof UnaryExpr u) {
            double inner = evaluate(u.getOperand());
            return u.getOperator() == '-' ? -inner : inner;
        }
        if (expr instanceof BinaryExpr b) return OperatorTable.apply(b.getOperator(), evaluate(b.getLeft()), evaluate(b.getRight()));
        if (expr instanceof CallExpr c) return registry.invoke(c.getFunction(), c.getArguments().stream().mapToDouble(this::evaluate).toArray());
        throw new EvalException("cannot evaluate " + expr.describe());
    }
}

File: fixtures/scale/dart/medium/lib/eval/evaluator.dart
import '../ast/assign_expr.dart';
import '../ast/binary_expr.dart';
import '../ast/call_expr.dart';
import '../ast/expr.dart';
import '../ast/number_expr.dart';
import '../ast/unary_expr.dart';
import '../ast/variable_expr.dart';
import '../parser/operator_table.dart';
import 'environment.dart';
import 'function_registry.dart';

class Evaluator {
  final Environment environment;
  final FunctionRegistry registry;

  Evaluator(this.environment, this.registry);

  double evaluate(Expr expr) {
    if (expr is NumberExpr) return expr.value;
    if (expr is VariableExpr) return environment.lookup(expr.name);
    if (expr is AssignExpr) {
      final value = evaluate(expr.value);
      environment.define(expr.name, value);
      return value;
    }
    if (expr is UnaryExpr) {
      final inner = evaluate(expr.operand);
      return expr.operator == '-' ? -inner : inner;
    }
    if (expr is BinaryExpr) {
      return applyOperator(expr.operator, evaluate(expr.left), evaluate(expr.right));
    }
    if (expr is CallExpr) {
      return registry.invoke(expr.function, expr.arguments.map(evaluate).toList());
    }
    throw StateError('cannot evaluate ${expr.describe()}');
  }
}
